Fixes bad camera positions and change options WMTS

gchoqueux opened the pull request:
Fixes camera bad positions when you use :

* setCameraTargetGeoPosition() #255
* setRange()
* pan() see #260
* createSceneGlobe()


Move WMTS's options in layer.options 

Add little doc addImageryLayer and addElevationLayer

**Warning, this PR doesn't fix incorrect position with setZoomLevel and setZoomScale**

gchoqueux committed:
refacto (WMTS) : move WMTS's options in layer.options

gchoqueux committed:
Fix (Camera): Return a promise in setRange

gchoqueux committed:
fix (GlobeControls) : fix error in pan() and with helperViewer

* fix bad up vector camera when pan is used
* hide helperViewer when picking position to avoid wrong depth testing

peppsac commented:
Please move the test commit out of this PR (big topic so it deserves its own PR).

nosy-b commented:
I just started reviewing with method` ApiGlobe.getCenter()`
In index.html it seems the initial position of camera is never properly used but is rounded up. 
Try for example to start centerd on Paris` const positionOnGlobe = { longitude: 2.351323, latitude: 48.856712, altitude: 25000000 };` 
Then ` ApiGlobe.getCenter()` will return {longitude: 2.3513230022207057, latitude: 49.00976271512532} which seems to be the real position as zooming get you north of Paris.

elias75015 commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -67,30 +67,32 @@ WMTS_Provider.prototype.preprocessDataLayer = function preprocessDataLayer(layer
-            min: minZoom,
-            max: maxZoom,
-        };
+            layer.zoom = {
```
Shouldn't it be in "layer.options" ?

gchoqueux replied:
ok, I move in options

elias75015 commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -634,15 +642,26 @@ ApiGlobe.prototype.setCenter = function setCenter(coordinates, isAnimated) {
+            if (pPosition.level) {
+                pPosition.range = this.getRangeFromZoomLevel(pPosition.level);
+            } else if (pPosition.scale) {
+                pPosition.range = this.getRangeFromScale(pPosition.scale);
```
- Only with the param "isAnimated" set to false : When the scale parameter is given, the view is tilted to 90°, even without the parameter "tilt" given to the function.
- I think it is more explicit to give the scale param as a fraction denominator than in decimal (ex : 500 instead of 0.002)

gchoqueux replied:
> I think it is more explicit to give the scale param as a fraction denominator than in decimal (ex : 500 instead of 0.002)

The scale of a map is the ratio,  we must be find a another name to this parameter

gchoqueux replied:
> Only with the param "isAnimated" set to false : When the scale parameter is given, the view is tilted to 90°, even without the parameter "tilt" given to the function.

Try with the new commit ` Fix (Viewer) :  add default value for setCenterAdvanced() `

elias75015 commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -634,15 +642,26 @@ ApiGlobe.prototype.setCenter = function setCenter(coordinates, isAnimated) {
-        return p;
+        if (pPosition.range === undefined) {
+            if (pPosition.level) {
+                pPosition.range = this.getRangeFromZoomLevel(pPosition.level);
```
The function takes well into account the level param (the view seems to be correctly positioned), but the getZoomLevel doesn't return the same value as the one given in the setCenterAdvenced function.

gchoqueux replied:
getZoomLevel will be fixed in #279

elias75015 commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -428,7 +428,9 @@ ApiGlobe.prototype.getCameraLocation = function getCameraLocation() {
-    return this.projection.cartesianToGeo(controlCam.getTargetCameraPosition());
+    const tg = controlCam.getTargetCameraPosition();
+    const coordinates = new Coordinates('EPSG:4978', tg.x, tg.y, tg.z).as('EPSG:4326');
+    return { longitude: coordinates.longitude(UNIT.DEGREE), latitude: coordinates.latitude(UNIT.DEGREE) };
```
Not sure about the length of the returned values (does it make sense ?) : 
latitude : 47.995276640947594
longitude : 1.0015039089052769

gchoqueux replied:
Yes maybe, especially since I do not know the precision of my calculation

elias75015 commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -660,7 +679,7 @@ ApiGlobe.prototype.setRange = function setRange(pRange, isAnimated) {
-        this.setSceneLoaded().then(() => {
+        return this.setSceneLoaded().then(() => {
             this.scene.currentControls().updateCameraTransformation();
             this.viewerDiv.dispatchEvent(eventRange);
```
If I do a setRange(10000), the view position is OK, but the getRange() does not return 10000.

The getRange seems to return old range value. Ex :
- setRange(10000) --> getRange() = 25011386.530042846
- setRange(200000) --> getRange() =10013.066997543165 (almost 10000)

gchoqueux replied:
try with `then` promise,  you do getRange() just after setRange(), While the setRange is not yet completed 

setRange(10000).then(() => getRange())

gchoqueux commented:
> Then ApiGlobe.getCenter() will return {longitude: 2.3513230022207057, latitude: 49.00976271512532} which seems to be the real position as zooming get you north of Paris.

Fix in last commit, It was due to an ellipsoidal/sphere approximation

gchoqueux committed:
doc (protocol) : Imagery and Elevation

gchoqueux committed:
fix (Camera) : fix setCameraTargetGeoPosition and setCameraTargetGeoPositionAdvanced

* fixes setCameraTargetGeoPositionAdvanced()
   - the parameters scale and level are taken into account
   - add default value for range, tilt and heading
   - add documentation

* fix setCameraTargetGeoPosition()
  - Add compensation of the altitude from the approximation of the ellipsoid by a sphere

gchoqueux committed:
Fix (GlobeControls): Correct initial target position

nosy-b commented:
LGTM

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -260,13 +303,13 @@ ApiGlobe.prototype.addElevationLayersFromJSONArray = function addElevationLayers
 ApiGlobe.prototype.getMinZoomLevel = function getMinZoomLevel(index) {
     var layer = this.getImageryLayers()[index];
     if (layer && layer.zoom) {
```
I guess it should be `layer.options.zoom`.

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -260,13 +303,13 @@ ApiGlobe.prototype.addElevationLayersFromJSONArray = function addElevationLayers
         var layers = this.getImageryLayers();
         let min = Infinity;
         for (var i = layers.length - 1; i >= 0; i--) {
             if (layers[i].zoom) {
```
same here.

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -628,7 +672,14 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
+            .as('EPSG:4978').xyz();
+        const diff = position3D.length() - oldPosition3D.length();
+        position3D.range = coordinates.range + diff;
+    }
```
Suggestion: put this piece of code in GlobeControls.

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -643,15 +694,37 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
+ * @param {Position} position
+ * @param {Number}  position.longitude  Coordinate longitude WGS84 in degree
+ * @param {Number}  position.latitude  Coordinate latitude WGS84 in degree
+ * @param {Number}  [position.tilt]  Camera tilt in degree
```
should not be surrounded by `[]`, right?

gchoqueux replied:
position.tilt is optionnal

vcoindet replied:
I agree with @gchoqueux

autra replied:
Yes yes, me too :-) Thanks for the precision

autra reviewed (commented):
Leaving some comments, as there are still some `layer.zoom` in the code. You might want to use `git grep` to make sure nothing is left.

Btw, why not keeping `layer.zoom` and not using `layer.options.zoom`? Do we really need the additional level (as more or less everything are layer options anyway)? 

Thanks

peppsac reviewed (dismissed):
Regardless of the changes themselves (no time for review atm) this PR has several problems:
- no PR summary (and the title is vague)
- 7 commits with broad topics
- commits messages are either empty or meaningless (e.g : `Now setRange() return Promise`)

gchoqueux commented:
> layer.options.zoom

zoom is specific to WMTS, so I moved zoom in `wmts_options` (see doc)

but this is not perfect, `attribution` should out of options

elias75015 commented on src/Core/Commander/Providers/WMTS_Provider.js:
```diff
@@ -241,9 +243,9 @@ WMTS_Provider.prototype.computeLevelToDownload = function computeLevelToDownload
 
     return Math.min(
-        layer.zoom.max,
+        layer.options.zoom.max,
```
A lot of 404 CORS errors are launched errors when we zoom beyond the zoom.max given as parameter of the layer.

elias75015 commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -643,15 +694,37 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
+ * @param {boolean}  isAnimated  Indicates if animated
+ * @return {Promise}
+ */
+ApiGlobe.prototype.setCameraTargetGeoPositionAdvanced = function setCameraTargetGeoPositionAdvanced(position, isAnimated) {
```
Not totally working for the moment (tested in web dev console) :
- range parameter is strangely handled (value given doesn't fit with the view)
- heading and tilt parameters not taken into account

gchoqueux commented:
@autra @peppsac  : remarks take into account

gchoqueux commented:
@elias75015 Branch is updated

gchoqueux commented:
@peppsac when will you be available?

peppsac commented:
@gchoqueux not this week, maybe next one.

gchoqueux commented:
@autra @peppsac, this PR needs reviewing

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -16,7 +16,8 @@ import Fetcher from '../../Providers/Fetcher';
 var sceneIsLoaded = false;
-var eventLoaded = new CustomEvent('globe-loaded');
+var eventLoaded = new CustomEvent('dataloaded');
+var eventInitialized = new CustomEvent('initialized');
```
Create events where you dispatch them. They can't be used several time safely

autra replied:
Actually, for all the events, but particularly for the events you modify in this PR: the only things it makes sense to put in `const` (not `var` btw) are the string itself:

```js
const DATA_LOADED_EVENT = 'dataloaded';

/* later */

this.dispatchEvent(new CustomEvent(DATA_LOADED_EVENT));

/* elsewhere */

itowns.viewer.addEventListener(itowns.viewer.DATA_LOADED_EVENT, /* ... */);
```

Webpack will then warn if there is a typo.

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
+ * The protocol rules wich parameters are then needed for the function.
  * @constructor
- * @param {Layer} layer.
+ * @param {Layer} layer
```
Missing the `Layer` typedef

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
+
+/**
+ * Options to wms protocol
+ * @typedef {Object} options_wms
```
By tradition, types are in CamelCase (and variables are also camelCase in itowns anyway).

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
- * This function adds an imagery layer to the scene. The layer id must be unique. The protocol rules wich parameters are then needed for the function.
+ * The intellectual property rights
+ * @typedef {Object} Attribution
+ * @property {String} name
```
the type should be the same string returned by the `typeof` operator, so `string` and not `String`

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -97,10 +97,49 @@ ApiGlobe.prototype.addGeometryLayer = function addGeometryLayer(layer) {
+ * @property {String} mimetype
+ * @property {String} tileMatrixSet
+ * @property {Array.<Object>} tileMatrixSetLimits The limits for the tile matrix set
+ * @property {Number} tileMatrixSetLimits.minTileRow Minimum row for tiles at the level
```
`number`

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -643,15 +689,37 @@ ApiGlobe.prototype.setCameraTargetGeoPosition = function setCameraTargetGeoPosit
-        const p = this.scene.currentControls().setOrbitalPosition(undefined, pPosition.heading, pPosition.tilt, isAnimated);
-        return p;
+    return this.setCameraTargetGeoPosition(position, isAnimated).then(() => {
+        if (position.range === undefined) {
```
`if (position.range)`

gchoqueux replied:
It is necessary to distinguish range equal zero and undefined

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -478,18 +478,27 @@ function GlobeControls(camera, domElement, engine) {
             return position;
         };
-    }());
+    }(this.pickingHelper));
```
Why do you need this closure here? `this.pickingHelper` won't change, I don't see the point of capturing engineGFX, and position is not really an internal state here, as you never read from it.

gchoqueux replied:
Yes indeed

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -478,18 +478,27 @@ function GlobeControls(camera, domElement, engine) {
+            if (enableTargetHelper) {
+                pickingHelper.visible = true;
+                cameraTargetOnGlobe.visible = true;
+            }
```
Can you explain why you need these 2 `if` here? It looks odd to have to do that.

gchoqueux replied:
if, there are targets helper, we must hide these targets for the picking

gchoqueux replied:
if there are  targets helper, we hide targets helper before picking
and after picking if there are  targets helper we display targets helper for rendering

gchoqueux replied:
The use of THREE.js layers in PR @239 will simplify this point

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -669,6 +679,9 @@ function GlobeControls(camera, domElement, engine) {
         state = CONTROL_STATE.NONE;
         lastRotation = [];
+        if (enableTargetHelper) {
+            this.dispatchEvent(this.changeEvent);
```
Note: typically why I'm pretty sure the dependency between c3DEngine and Controls should not exist.

gchoqueux replied:
Yes it is removed in PR #239

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1122,13 +1135,15 @@ function GlobeControls(camera, domElement, engine) {
         state = controlState || CONTROL_STATE.ORBIT;
         update();
-        updateCameraTargetOnGlobe.bind(this)();
+        if (forceNoUpdateTarget === undefined || forceNoUpdateTarget) {
```
Please use positive logic : `forceUpdate`, `preventUpdate`, whatever you prefer. Negative naming is difficult to read and error-prone. Indeed, here when forceNoUpdateTarget is true, we *do* update the target on globe, which seems contradictory to me.

Btw, why do you need to bypass the update of the camera in some case ?

gchoqueux replied:
Because the target must wait for the data to be loaded (in particular the relief data)

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1269,6 +1283,15 @@ GlobeControls.prototype.getCameraTargetPosition = function getCameraTargetPositi
     const center = this.getCameraTargetPosition();
 
+    if (position.range) {
+        // Compensation of the altitude from the approximation of the ellipsoid by a sphere
```
Then the question is: where do we approximate the ellipsoid by a sphere that in turn make this hack necessary?

autra commented on src/Renderer/c3DEngine.js:
```diff
@@ -15,7 +15,7 @@ import RendererConstant from './RendererConstant';
 var instance3DEngine = null;
 
-function c3DEngine(scene, positionCamera, viewerDiv, debugMode, gLDebug) {
+function c3DEngine(scene, positionCamera, targetCamera, viewerDiv, debugMode, gLDebug) {
```
This is specific to GlobeControls, so to a Globe. I would prefer not to add that as a parameter to c3DEngine.

gchoqueux replied:
Yes, positionCamera and targetCamera are specific GlobeControl
There are removed in PR #239

autra replied:
Where?

autra commented on src/Scene/Scene.js:
```diff
@@ -39,12 +39,17 @@ const RENDERING_ACTIVE = 1;
  * - Scene (and subobjects) should be instanciable several times.
  */
-function Scene(crs, positionCamera, viewerDiv, debugMode, gLDebug) {
+function Scene(crs, coordinate, viewerDiv, debugMode, gLDebug) {
```
Why this change? you don't need it and it's less explicit for the caller.

autra commented on src/Scene/Scene.js:
```diff
@@ -39,12 +39,17 @@ const RENDERING_ACTIVE = 1;
 
+    const positionCamera = coordinate.as(crs);
+    const coordinateTargetCamera = coordinate.clone();
+    coordinateTargetCamera._values[2] = 0;
```
undescored properties are meant to be considered as private. You can't use it here. You either need to move this code or to add a function to Coordinate (`fromLonLatHeight` maybe)

gchoqueux commented on src/Main.js:
```diff
@@ -4,13 +4,14 @@
     viewer: new ApiGlobe(),
 };
 scope.itowns = itowns;
+itowns.viewer.INITIALIZED_EVENT = INITIALIZED_EVENT;
```
@autra how export INITIALIZED_EVENT?

autra replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export

autra replied:
And you want to have a look at Main.js

gchoqueux commented:
@autra PR updated

gchoqueux commented:
@autra PR updated

autra commented:
@gchoqueux still some unanswered comments (https://github.com/iTowns/itowns2/pull/273/files/eaea9f8672ed35e22a10deca40bb8bb1b3a5b813#diff-df61a67eee71572718ef066564a3a0af). Thanks.

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -322,13 +367,9 @@ ApiGlobe.prototype.createSceneGlobe = function createSceneGlobe(coordCarto, view
-            viewerDiv.dispatchEvent(eventLoaded);
-        } else {
-            viewerDiv.dispatchEvent(eventError);
+            viewerDiv.dispatchEvent(new CustomEvent(DATA_LOADED_EVENT));
```
The event DATA_LOADED_EVENT is only listened to in this function. So you can remove it and put the 2 calls in the listener here. We don't need to add an event.

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -1122,13 +1131,15 @@ function GlobeControls(camera, domElement, engine) {
         state = controlState || CONTROL_STATE.ORBIT;
         update();
-        updateCameraTargetOnGlobe.bind(this)();
+        if (preventUpdate === undefined || preventUpdate) {
```
If preventUpdate is true, `updateCameraTargetOnGlobe` is called, which seems contradictory to me.

Did you want to 
```js
if (!preventUpdate) { // 
   // update
}
```

Or even better, do you know javascript [now has default parameters value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters).

I then suggest to have the following signature: 
```js
function updateCameraTransformation(controlState, updateCameraTarget = true)
```

 and simply use `if (updateCameraTarget)` here.

autra commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -478,18 +479,22 @@ function GlobeControls(camera, domElement, engine) {
+        if (enableTargetHelper) {
+            pickingHelper.visible = true;
+            cameraTargetOnGlobe.visible = true;
+        }
```
Can you elaborate on these 2 ifs? I see no difference when I remove them. The `targetHelpers` are the axis on the globe where you clicked and where the target is, right? Picking is done with s+click right?

gchoqueux replied:
if 'enableTargetHelper' is enable there's error in picking position because helpers write in depth buffer. Picking position is do every time you click on globe.  s+click right is for selection.

> The targetHelpers are the axis on the globe where you clicked and where the target is, right

yes

autra reviewed (changes requested):
2 last remarks, one question, and we're good to go.

Before I merge, please split your review commits, and squash them in the relevant commits.

gchoqueux committed:
Fix (Camera): fix target position update after setRange()

  * Remove 'globe-loaded' event
  * Add initialized event : when iTowns is initialized (called one time)

gchoqueux commented:
updated

autra commented:
LGTM. A few nitpicks on commit messages with proposals:

>  fix (Camera) : fix setRange() doesn't return Promise
> 
>  return Promise in setRange()

Change to:
> Fix (Camera): Return a promise in setRange

In third commit:

> hide helperViewer when pick position

to 
> hide helperViewer when picking position to avoid wrong depth testing

In fifth:

> fix (GlobeControls) : fix initial target position is incorrect

to 
> Fix (GlobeControls): Correct initial target position

In sixth: there is a obsolete explanation about data-loaded (removed), and change:

> Fix (Camera): fix update target position after setRange()

to 
> Fix (Camera): fix target position update after setRange()

(if I understand well what you meant)

gchoqueux commented:
commits message updated

autra commented:
Thanks!
